lib/DxilPIXPasses/DxilPIXAddTidToAmplificationShaderPayload.cpp

```diff
@@ -117,29 +117,51 @@ bool DxilPIXAddTidToAmplificationShaderPayload::runOnModule(Module &M) {
 
     llvm::IRBuilder<> B(UserInstruction);
 
-    auto ThreadIdFunc =
-        HlslOP->GetOpFunc(DXIL::OpCode::ThreadId, Type::getInt32Ty(Ctx));
-    Constant *Opcode = HlslOP->GetU32Const((unsigned)DXIL::OpCode::ThreadId);
     Constant *Zero32Arg = HlslOP->GetU32Const(0);
     Constant *One32Arg = HlslOP->GetU32Const(1);
     Constant *Two32Arg = HlslOP->GetU32Const(2);
 
-    auto ThreadIdX =
-        B.CreateCall(ThreadIdFunc, {Opcode, Zero32Arg}, "ThreadIdX");
-    auto ThreadIdY =
-        B.CreateCall(ThreadIdFunc, {Opcode, One32Arg}, "ThreadIdY");
-    auto ThreadIdZ =
-        B.CreateCall(ThreadIdFunc, {Opcode, Two32Arg}, "ThreadIdZ");
-
-    auto *XxY =
-        B.CreateMul(ThreadIdX, HlslOP->GetU32Const(m_DispatchArgumentY));
-    auto *XplusY = B.CreateAdd(ThreadIdY, XxY);
-    auto *XYxZ = B.CreateMul(XplusY, HlslOP->GetU32Const(m_DispatchArgumentZ));
-    auto *XYZ = B.CreateAdd(ThreadIdZ, XYxZ);
+    auto GroupIdFunc =
+        HlslOP->GetOpFunc(DXIL::OpCode::GroupId, Type::getInt32Ty(Ctx));
+    Constant *GroupIdOpcode =
+        HlslOP->GetU32Const((unsigned)DXIL::OpCode::GroupId);
+    auto *GroupIdX =
+        B.CreateCall(GroupIdFunc, {GroupIdOpcode, Zero32Arg}, "GroupIdX");
+    auto *GroupIdY =
+        B.CreateCall(GroupIdFunc, {GroupIdOpcode, One32Arg}, "GroupIdY");
+    auto *GroupIdZ =
+        B.CreateCall(GroupIdFunc, {GroupIdOpcode, Two32Arg}, "GroupIdZ");
+
+    auto *GroupYxNumZ = B.CreateMul(
```

Comment:
Following this code could benefit from a comment with the formula being constructed.
I would have expected something like: `gid.X + gid.Y * DimX + gid.Z * (DimX * DimY)`
But this looks like: `gid.X * (DimY * DimZ) + gid.Z + gid.Y * DimZ`
That just makes it flatten in a different way.  Instead of minor-to-major dims being X, Y, Z, they are: Z, Y, X.

Is that the intent?

Reply:
That is the intent. Yes, this definitely needs some comments. Added. Thanks!